File: 20. Valid Parentheses/using_hashmap_stack_1.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> hashmap = new HashMap<Character, Character>() {{
            put('{', '}');
            put('(', ')');
            put('[', ']');
        }};
        Stack<Character> st = new Stack<>();
        for (char bracket: s.toCharArray()) {
            if (hashmap.containsKey(bracket)){
                st.push(bracket);
            } else {
                if (st.empty() || hashmap.get(st.pop()) != bracket) {
                    return false;
                }
            }
        }
        return st.empty();
    }
}

File: 20. Valid Parentheses/using_hashmap_stack_1.cpp
#include <stack>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> hashmap = {{'{','}'}, {'(',')'}, {'[',']'}};
        stack<char> st;
        for (char bracket: s){
            if (hashmap.count(bracket)){
                st.push(bracket);
            }
            else {
                if (st.empty() || bracket != hashmap[st.top()]) {
                    return false;
                }
                st.pop();
            }
        }
        return st.empty();   
    }
};